### PR TITLE
Rank Peers Before Disconnecting them

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Addition.

**What does this PR do? Why is it needed?**

Feature Improvement.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/p2p/connection_gater.go [hunk 1]
```diff
@@ -13,11 +13,17 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// limit for rate limiter when processing new inbound dials.
-const ipLimit = 4
+const (
+	// limit for rate limiter when processing new inbound dials.
```
Comment (farazdagi) [inline beacon-chain/p2p/connection_gater.go:17]:
```suggestion
	// Limit for rate limiter when processing new inbound dials.
```

File: beacon-chain/p2p/connection_gater.go [hunk 2]
```diff
@@ -13,11 +13,17 @@ import (
...
+	ipLimit = 4
 
-// burst limit for inbound dials.
-const ipBurst = 8
+	// burst limit for inbound dials.
```
Comment (farazdagi) [inline beacon-chain/p2p/connection_gater.go:20]:
```suggestion
	// Burst limit for inbound dials.
```

File: beacon-chain/p2p/connection_gater.go [hunk 3]
```diff
@@ -13,11 +13,17 @@ import (
...
+	ipBurst = 8
+
+	// high watermark buffer signifies the buffer till which
```
Comment (farazdagi) [inline beacon-chain/p2p/connection_gater.go:23]:
```suggestion
	// High watermark buffer signifies the buffer till which
```

File: beacon-chain/p2p/connection_gater.go [hunk 4]
```diff
@@ -44,7 +50,7 @@ func (s *Service) InterceptAccept(n network.ConnMultiaddrs) (allow bool) {
 			"reason": "exceeded dial limit"}).Trace("Not accepting inbound dial from ip address")
 		return false
 	}
-	if s.isPeerAtLimit() {
+	if s.isPeerAtLimit(true) {
```
Comment (farazdagi) [inline beacon-chain/p2p/connection_gater.go:53]:
```suggestion
	if s.isPeerAtLimit(true /* inbound */) {
```

File: beacon-chain/p2p/discovery.go [hunk 5]
```diff
@@ -68,7 +68,7 @@ func (s *Service) listenForNewNodes() {
 		if s.ctx.Err() != nil {
 			break
 		}
-		if s.isPeerAtLimit() {
+		if s.isPeerAtLimit(false) {
```
Comment (farazdagi) [inline beacon-chain/p2p/discovery.go:71]:
```suggestion
		if s.isPeerAtLimit(false /* inbound */) {
```

File: beacon-chain/p2p/peers/status.go [hunk 6]
```diff
@@ -120,8 +123,9 @@ func (p *Status) Add(record *enr.Record, pid peer.ID, address ma.Multiaddr, dire
 		return
 	}
 	peerData := &peerdata.PeerData{
-		Address:   address,
-		Direction: direction,
+		Address:        address,
+		Direction:      direction,
+		FirstAddedTime: timeutils.Now(),
```
Comment (farazdagi) [inline beacon-chain/p2p/peers/status.go:128]: Do we also need to make sure that `store.PeerDataGetOrCreate()` method sets `FirstAddedTime` to `timeutils.Now()`?

**Review by farazdagi — COMMENTED** [review]
Looks good, just couple of minor things.

**Comment by ibrahim419** [comment]
Clm

**Comment by ibrahim419** [comment]
Chery

**Pull request closed without merging**